From 7c9a8430cdf0082410bd5ab19ba69e7b53c0dbdf Mon Sep 17 00:00:00 2001
From: Jan Sondhauss <jan.sondhauss@wago.com>
Date: Mon, 17 Feb 2020 14:11:42 +0100
Subject: [PATCH] drivers:net:dsa: mv88e6xxx: make port led register
 configurable via of-tree

With the 750-8211 it is necessary to configure the led-ctrl differently for
the ports with the internal phys and the port connected to the SFP modules.

The marvel mv88e6321 supports up to 6 leds, with the leds being connected to
ports 3,4 and 5. For the pfc 8211 the led-ctrl of port 5 is configured to
show the activity of port 0 and 1.

Signed-off-by: Jan Sondhauss <jan.sondhauss@wago.com>
---
 arch/arm/boot/dts/am335x-pfc-750_8211.dts |  2 ++
 drivers/net/dsa/mv88e6xxx/chip.c          | 22 +++++++++++++++++++---
 2 files changed, 21 insertions(+), 3 deletions(-)

diff --git a/arch/arm/boot/dts/am335x-pfc-750_8211.dts b/arch/arm/boot/dts/am335x-pfc-750_8211.dts
index c9417d2..cfc7afc 100644
--- a/arch/arm/boot/dts/am335x-pfc-750_8211.dts
+++ b/arch/arm/boot/dts/am335x-pfc-750_8211.dts
@@ -105,6 +105,8 @@
 			   0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF
 			   0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF>;
 
+	led-ctrl = /bits/ 16 <0x0000 0x0000 0x0000 0x0088 0x0088 0x0000 0x0000 0x0000>;
+
 	ports {
 		port@0 {
 			reg = <0>;
diff --git a/drivers/net/dsa/mv88e6xxx/chip.c b/drivers/net/dsa/mv88e6xxx/chip.c
index 6e1c642..06ea906 100644
--- a/drivers/net/dsa/mv88e6xxx/chip.c
+++ b/drivers/net/dsa/mv88e6xxx/chip.c
@@ -2843,6 +2843,22 @@ static bool mv88e6xxx_has_ext_phy(struct dsa_switch *ds, int port)
 	return of_property_read_bool(np, "phy-external");
 }
 
+static u16 mv88e6xxx_of_get_led_ctrl(struct mv88e6xxx_chip *chip, int port,
+				     u16 def_value)
+{
+	u16 led_ctrl[DSA_MAX_PORTS];
+	struct device_node *np = chip->mdio_np;
+
+	if (WARN_ON(mv88e6xxx_num_ports(chip) > DSA_MAX_PORTS))
+		return def_value;
+
+	if (!of_property_read_u16_array(np, "led-ctrl", led_ctrl,
+					mv88e6xxx_num_ports(chip)))
+		return led_ctrl[port];
+
+	return def_value;
+}
+
 static int mv88e6xxx_setup_port(struct mv88e6xxx_chip *chip, int port)
 {
 	struct dsa_switch *ds = chip->ds;
@@ -2850,9 +2866,9 @@ static int mv88e6xxx_setup_port(struct mv88e6xxx_chip *chip, int port)
 	u16 reg;
 
 	reg = PORT_LED_CTRL_UPDATE
-		| PORT_LED_CTRL_LED01
-		| LED0_SELECT_10XX_CONT
-		| LED1_SELECT_ACT;
+		| mv88e6xxx_of_get_led_ctrl(chip, port, PORT_LED_CTRL_LED01
+					  | LED0_SELECT_10XX_CONT
+					  | LED1_SELECT_ACT);
 
 	mv88e6xxx_port_write(chip, port, PORT_LED_CONTROL, reg);
 
-- 
2.7.4

